### PR TITLE
Fix branch invariants

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

I discovered when working on some other workflows that editing `terms`, `types`, and `children` could cause the `isEmpty0` cached state on a branch0 to become out of date, which then broke future `deepTerms` derivations.

This fixes the updates inside those lenses, and also no longer exports those internal fields, since it's risky to let callers update them without thinking about the invariants. Probably should've done this a while ago tbh.

Note: Most interesting changes are in [THIS COMMIT](https://github.com/unisonweb/unison/pull/5027/commits/0dcbc126a1084be813a7491165abfd6ce7a519a0#diff-e3f9cef9eeca485df68a37548d19eb990f7d18427af3bcaa35c084e8cb3f1efc)
A lot of the rest is noise from re-writing field getters to the lensy accessors and tweaks from the new prelude exports.

## Implementation notes

* Don't export any Branch0 fields, as updating them manually can break invariants
* Fix all usage sites to instead use the lensy accessors, which maintain invariants
* Added some basic lensy stuff to the Prelude since we use it all over nowadays.

## Test coverage

This one is proving tricky to test, AFAICT it only causes problems when:

1. You have a branch that has some sub-namespace that is empty, but DOES have history (so it exists in the in-memory branch)
2. You add some terms to a *child* of that branch using `slurpAdds`. This uses the `terms` lens, which on trunk does NOT update the sub-namespace's `isEmpty0` field, resulting in a namespace that's reported as _empty_, but whose children DOES have terms.
3. You then pass exactly that in-memory branch to `diffNamespace` directly (without reloading it from the codebase, which would recalculate `isEmpty0`). `diffNamespace` crawls through nonEmpty children, and will _skip_ the branch with the new terms.

It's hard to test because I'm not aware of a way to set up this situation using only existing UCM commands, I only ran into it on my new `commit` command branch.

It's also very arduous to construct/manipulate terms outside of transcripts in unit tests 🤔 

I guess if it's something that can't be tested via UCM maybe it's okay to wait until that branch adds some tests for the new command? Open to ideas.

### CODE REVIEW

File: parser-typechecker/src/Unison/Codebase/Branch/Type.hs [hunk 1]
```diff
@@ -88,4 +113,199 @@ history :: Iso' (Branch m) (UnwrappedBranch m)
...
+    )
+
+terms :: Lens' (Branch0 m) (Star Referent NameSegment)
+terms =
+  lens
+    _terms
+    \branch terms ->
+      branch {_terms = terms}
+        & deriveDeepTerms
+        & deriveIsEmpty
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Branch/Type.hs:131]: Here's where we were forgetting to re-derive `isEmpty`. (see same on edits, types, and children)

File: parser-typechecker/src/Unison/Codebase/Branch/Type.hs
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Branch/Type.hs:1]: Changes to this module:

* Branch0 fields/constructor are no longer exported
* I copied over the lensy accessors from `Branch.hs` to here to export as an external interface
* Rederive 'isEmpty' in the appropriate lensy accessors.

Everything else is mostly just copied over from `Branch.hs` without changing.

File: parser-typechecker/src/Unison/Codebase/Branch.hs [hunk 2]
```diff
@@ -5,7 +5,7 @@ module Unison.Codebase.Branch
   ( -- * Branch types
     Branch (..),
     UnwrappedBranch,
-    Branch0 (..),
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Branch.hs:8]: No longer export the record fields.

File: parser-typechecker/src/Unison/Codebase/Branch.hs [hunk 3]
```diff
@@ -198,179 +204,13 @@ deepTermReferences =
 deepTypeReferences :: Branch0 m -> Set TypeReference
 deepTypeReferences = R.dom . deepTypes
 
-terms :: Lens' (Branch0 m) (Star Referent NameSegment)
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Branch.hs:201]: Move all these accessors into the `Branch.Type` module

**Review by aryairani — APPROVED** [review]
Thanks — go ahead and merge once it's mergeable again, @ChrisPenner

**Pull request merged**